csrc/multidevice/communication.h

```diff
@@ -68,16 +69,35 @@ class Communication {
       Communicator& comm,
       std::optional<CommunicatorBackend> backend = std::nullopt) = 0;
 
+  // Allocates a contiguous intermediate tensor.
+  // Required for NCCL/UCC which expect contiguous input/output buffers.
+  // input tensor was sharded like [a b .. DIDx ... c] where DIDx is the axis
+  // parallelized on DIDx the output tensor elements will be ordered as
+  //  [DIDx a b c]  with device dimensions pushed to the outer most axis
+  virtual bool requiresRelayoutOutputTensor(
+      TensorView* input_tv,
+      TensorView* output_tv);
+  virtual at::Tensor allocateOutputTensor(TensorView* tv, at::Tensor& tensor);
+  virtual void relayoutOutputTensor();
+
  protected:
   // argument "name" is only used for printing
   // argument "has_root" indicates if the communication is rooted
   Communication(CommParams params, std::string name, bool has_root = true);
+  Communication(std::string name, bool has_root = true);
```

Comment:
Why do we need this constructor ?

Reply:
It seems that the other constructors with CommParams become obsolete with this patch.

Reply:
The CommParams constructor are used by the CommunicationTests. 

If I split the PR like you mentioned earlier, I can update the CommunicationTests and get rid of the obsolete constructor in the first PR.
The second PR will parallelize non-outmost axis in the multidevice Fusion and update the PipelineTests. wdyt?